---
title: "Reproducible and Interactive Research"
author: "Cory Merow & Adam Wilson"
date: "9/19/2017"
output:
  html_document:
    code_folding: show
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(cache=TRUE)
```

# Getting started

You may need to install some software for this sequence of 3 tutorials.

- **RStudio** (convienient way to interact with R) :<br> https://www.rstudio.com/products/rstudio/download/#download 
- **git** (version control): <br> https://git-scm.com/downloads
- Create a **GitHub** account (sharing code online) at <br> [https://github.com/](https://github.com/)
- **R packages** (for these demos):<br> `install.packages(c('knitr','shiny','dplyr','ggplot2','maps','spocc','wallace','rmarkdown'), dep=T)`
- **pandoc** (to convert among document formats like html, pdf, docx): <br>http://pandoc.org/installing.html
<!-- ----------------------------------------------------------------------------- -->
<!-- ----------------------------------------------------------------------------- -->

<br>

***

<br>

# Overview

- Research Reports with R markdown
- Version control with Github
- Interactive presentations
- Developing Web Applications with Shiny
- Developing R websites (like this)

The same set of tools is used for all these!

<br>

## Benefits are straightforward

- **Verification & Reliability**: Find and fix bugs. 
- **Transparency**: Increased citation count, broader impact, institutional memory
- **Efficiency**: Reduces duplication of effort, Colleagues like you and ask for advice
- **Flexibility**: When you don’t 'point-and-click' you gain many new analytic options.

<br>

## But limitations are substantial

- Classified/sensitive/big data
- Nondisclosure agreements & intellectual property 
- Competition
- Few expect reproducibility 
- No uniform standards
- Inertia & embarassment

<br>

## A spectrum of reproducibility
<img src="07_assets/peng-spectrum.jpg" alt="alt text" width="800">

<small>Peng 2011, Science 334(6060) pp. 1226-1227</small>

<br>

## Code > Click trails {.columns-2}

- Clicks = Lots of human effort for tedious & time-wasting tasks
- Clicks = Error-prone due to manual & ad hoc data handling
- Clicks = Difficult to record -  hard to reconstruct a 'click history'
- Code = Improved transparency, automation, maintanability, accessibility, standardisation, modularity, portability, efficiency, communicability of process (what more could we want?)
- Code = Plain text files readable for a _long_ time
  
<!-- ## Literate programming: for and against   {.columns-2} -->

<!-- **Pros**-->

<!-- - Text & code in one place, in logical order -->
<!-- - Tables and figures automatically updated -->
<!-- - Automatic test when building document -->

<!-- **Cons** -->

<!-- - Text and code in one place; can be hard to read -->
<!-- - Can slow down the processing of documents (use caching!) -->

<br>

***

<br>

# Reproducible Research in R

- Programming language:  
    - The machine-readable part: R
- Document formatting language 
    - Markdown: lightweight document formatting syntax. Easy to write, read and publish as-is.
- Human-readable part
    - `rmarkdown`:  minor extensions to allow R code display and execution, embed images in html files, equations

`knitr`: Dynamic documents in R  {.columns-2}

- Narrative and code in the same file or explicitly linked
- When data or narrative are updated, the document is automatically updated

<!-- --------------------------------------------------------------------------------------- -->
<!-- --------------------------------------------------------------------------------------- -->
<br>

***

<br>

#  Demo
<br>

## R Markdown Cheatsheet

<a href="https://www.rstudio.com/wp-content/uploads/2015/02/rmarkdown-cheatsheet.pdf"> <img src="07_assets/rmarkdown.png" alt="alt text" width="700"></a>

<small><small><small>[https://www.rstudio.com/wp-content/uploads/2015/02/rmarkdown-cheatsheet.pdf](https://www.rstudio.com/wp-content/uploads/2015/02/rmarkdown-cheatsheet.pdf)</small></small></small>


The [reference guide](https://www.rstudio.com/wp-content/uploads/2015/03/rmarkdown-reference.pdf) is a little more comprehensive

<br>

## Create new file
In Rstudio:
**File -> New File -> RMarkdown -> Document -> HTML**

<img src="07_assets/rmarkdownwindow.png" alt="alt text" width="600">

This will open a document that looks like this:

<img src="07_assets/Example_knitr_Template.png" alt="alt text" width="750">

<!-- <img src="07_assets/Example_knitr_Template.png" alt="alt text" width="700"> -->

<br>

## Step 1: Load packages

All R code to be run must be in a _code chunk_ like this:
```{r, eval=F,asis=T}
#```{r,eval=F}
CODE HERE
#```
```

Add a new _code chunk_ at the bottom of this template file to load these packages (you may need to install some packages):

```{r, message=F}
library(dplyr)
library(ggplot2)
library(maps)
library(spocc)
```

> Do you think you should put `install.packages()` calls in your script?

<br>

## Step 2: Load data

Now use the `occ()` function in **new** _code chunk_ to download all the _occurrence_ records for the American robin (_Turdus migratorius_) from the [Global Biodiversity Information Facility](gbif.org).

<img src="07_assets/Turdus-migratorius-002.jpg" alt="alt text" width="200">

<small><small><small>Licensed under CC BY-SA 3.0 via [Wikimedia Commons](https://commons.wikimedia.org/wiki/File:Turdus-migratorius-002.jpg#/media/File:Turdus-migratorius-002.jpg)</small></small></small>

```{r, warning=F} 
## define which species to query
sp='Turdus migratorius'

## run the query and convert to data.frame()
d = occ(query=sp, from='ebird',limit = 100) %>% occ2df()
```
<br>

## Step 3: Map it

```{r, fig.width=6}
# Load coastline
map=map_data("world")

ggplot(d,aes(x=longitude,y=latitude))+
  geom_polygon(aes(x=long,y=lat,group=group,order=order),data=map)+
  geom_point(col="red")+
  coord_equal()
```

<br>

## Step 4: YAML header

Update the YAML header to keep the markdown file from this:
```{r, eval=F}
title: "Untitled"
author: "Adam M. Wilson"
date: "October 31, 2016"
output: html_document
```

To this (you need to save the markdown (md) document to make the html):
```{r, eval=F}
title: "Demo"
author: "Adam M. Wilson"
date: "October 31, 2016"
output: 
  html_document:
      keep_md: true
```

And click `knit HTML` to generate the output

<img src="07_assets/knitbutton.png" alt="alt text" width="600">

Changing this YAML header is all that's needed to change among document types. Try experimenting. (pdf won't work unless you have latex installed.)

<img src="07_assets/yaml.png" alt="alt text" width="600">

<br>

## Files

kniting  produces a number of files all extracted from the Rmd you wrote, seen below

<img src="07_assets/rmd_file_list.png" alt="alt text" width="600">

* .Rmd - you wrote this
* .md - plain markdown, with the R code already run
* .R - just the R code chunks, useful for rerunning just the code
* .html - the output file for others to view
* /YourDocumentName_files - any files generated from running the R code

In case you had trouble generating this document, [<i class="fa fa-file-code-o fa-3x" aria-hidden="true"></i> the R Script associated with this page is available here](07_assets/demo/Demo.Rmd). 

<br>

## Step 5:  Explore markdown functions

1. Use the Cheatsheet to add sections and some example narrative.  
2. Add more figures or different versions of a figure
3. Check out the `kable()` function for tables (e.g. `kable(head(d))`)
4. Some example of slightly more complex Rmd files:
  - [<i class="fa fa-file-code-o fa-3x" aria-hidden="true"></i> the vignette from the R package _bossMaps_](07_assets/bossMaps_demo_rmd.Rmd). 
  <!-- Additions to note:  -->
  <!--     * the _knitr::opts_chunk$set_ allows you to change some settings throughout the document -->
  <!--     * code chunk argument _message=F_ to suppress function messages -->
  <!--     * code chunk argument _fig.width_ to customize the size of a -->
* The [R markdown gallery](http://rmarkdown.rstudio.com/gallery.html) with examples of interactive documents, websites, books, and presentations, that we'll demonstrate in the next tutorials

<br>

## Step 6: Tips

* Caching will speed up compiling
```{r demosetup, as.is=T,eval=FALSE}
#```{r,eval=F}
knitr::opts_chunk$set(cache=TRUE)
#```
```
* Hide the code, just show the results
```{r demochunk1, as.is=T,eval=FALSE}
#```{r, echo=FALSE}
```
* Don't evaluate the code (maybe its slow or broken)
```{r demochunk2, as.is=T,eval=FALSE}
#```{r, eval=FALSE}
```
<!-- --------------------------------------------------------------------------------------- -->
<!-- --------------------------------------------------------------------------------------- -->
<br>

***

<br>

# More Background

## Key References

* [Rmarkdown homepage](http://rmarkdown.rstudio.com/)
* [html document format options](http://rmarkdown.rstudio.com/html_document_format.html) for the YAML header
* [using code chunks](http://rmarkdown.rstudio.com/authoring_rcodechunks.html)

<br>

## Pandoc: document converter  {.columns-2}

How does markdown magically convert between document types (html, pdf, docx)? 

A universal document converter, open source, cross-platform

* Write code and narrative in rmarkdown   
* knitr->markdown (with computation)   
* use pandoc to get HTML/PDF/DOCX
* its also very easy and useful from the command line: https://pandoc.org/demos.html


<img src="07_assets/pandoc-workflow-rmd-md.png" alt="alt text" width="100%">
<small><small><small>http://kieranhealy.org/blog/archives/2014/01/23/plain-text/ </small></small></small>

<br>

## Colophon

* [Slides based on Ben Marwick's presentation to the UW Center for Statistics and Social Sciences (12 March 2014)](https://github.com/benmarwick/CSSS-Primer-Reproducible-Research) ([OrcID](http://orcid.org/0000-0001-7879-4531))

<br>

## References

See Rmd file for full references and sources

```{r, echo=FALSE}
## Scholarly literature sources for this presentation

# Buckheit, J.B. and Donoho, D.L. Wavelab and reproducible research. (1995).
# Morin, A. et al. Shining light into black boxes. Science. 336, (2012), 159-160.
# King, G. Replication, Replication. PS: Political Science and Politics. (1995).
# Schofield, P.N. et al. Post-publication sharing of data and tools. Nature. 461, (2009), 171-173.
# Birney, E. et al. Prepublication data sharing. Nature. 461, (2009), 168-70.
# Peng, R.D. Reproducible research and Biostatistics. Biostatistics (Oxford, England). 10, (2009), 405-408.
# Vandewalle, P. et al. Reproducible research in signal processing - What, why, and how. IEEE Signal Processing Magazine.
# 26, (2009), 37-47.
# Stodden, V. The Legal Framework for Reproducible Scientic Research: Licensing and Copyright. Computing in Science &
# Engineering. 11, (2009), 35-40.
# V. Stodden, “Trust Your Science? Open Your Data and Code,” Amstat News, 1 July 2011; http://magazine.amstat.org/blog/2011/07/01/trust-your-science/
# Stodden V, Guo P, Ma Z (2013) Toward Reproducible Computational Research: An Empirical Analysis of Data and Code Policy Adoption by Journals. PLoS ONE 8(6): e67111. doi:10.1371/journal.pone.0067111
# Stodden V  2010 The Scientific Method in Practice: Reproducibility in the Computational Sciences. MIT Sloan School Working Paper 4773-10. http://ssrn.com/abstract=1550193
# Merali, Z. Error: Why scientic programming does not compute. Nature. (2010), 6-8.
# Barnes, N. Publish your computer code: it is good enough. Nature. 467, (2010), 753.
# LeVeque, R.J. Python tools for reproducible research on hyperbolic problems. Computing in Science & Engineering. (2009),
# 19-27. 
# LeVeque, R.J. Wave propagation software, computational science, and reproducible research. Proceedings of the International Congress of Mathematicians (Madrid, Spain, 2006), 1-27.
# LeVeque, R, Stodden, V., & Mitchell, I. (2012). Reproducible Research for Scientific Computing: Tools and Strategies for Changing the Culture. Computing in Science and Engineering, 14(4), 13–17
# Price, K. Anything You Can Do, I Can Do Better (No You Can't)... Computer Vision, Graphics, and Image Processing. (1986), 387-391.
# Piwowar, H. a et al. Sharing detailed research data is associated with increased citation rate. PloS one. 2, (2007), 308.
# Wilson, G. et al. Best Practices for Scientic Computing. 1-6.
# Drummond, C. Reproducible Research: a Dissenting Opinion. (2012), 1-10.
# Ioannidis, J.P. a et al. Repeatability of published microarray gene expression analyses. Nature genetics. 41, (2009), 149-55.
# Savage, C.J. and Vickers, A.J. Empirical study of data sharing by authors publishing in PLoS journals. PloS one. 4, (2009),
# 7078.
# Quirk, J. Computational Science \Same Old Silence, Same Old Mistakes" \Something More Is Needed..." Adaptive Mesh
# Reenement-Theory and Applications. (2005), 4-28.
# McCullough, B.D. Got Replicability? The Journal of Money, Credit and Banking Archive. Econ Journal Watch. 4, (2007),
# 326-337.
# McCullough, B.D. Do economics journal archives promote replicable research?. Economics Journal Archives. (2008).
# Manolescu, I. et al. Repeatability & Workability Evaluation of SIGMOD 2009. SIGMOD 2009 (2009), 2-4.
# Freire, J. et al. Computational reproducibility: state-of-the-art, challenges, and database research opportunities. SIGMOD
# 2012 (2012), 593-596.
# McCullough BD, Heiser DA (2008). “On the Accuracy of Statistical Procedures in Microsoft Excel 2007.” Computational Statistics & Data Analysis, 52, 4570–4578.
# Sandve GK, Nekrutenko A, Taylor J, Hovig E (2013) Ten Simple Rules for Reproducible Computational Research. PLoS Comput Biol 9(10): e1003285. doi:10.1371/journal.pcbi.1003285
# Gentleman, R. and Temple Lang, D. (2007). Statistical analyses and reproducible research. Journal of Computational and Graphical Statistics 16, 1–23.
# Leisch F, Eugster M and Hothorn T 2011. Executable Papers for the R Community: The R2 Platform for Reproducible Research. Procedia Computer Science 4(0), 618-626.
# Rossini, Anthony and Leisch, Friedrich, "Literate Statistical Practice" (March 2003). UW Biostatistics Working Paper Series. Working Paper 194. http://biostats.bepress.com/uwbiostat/paper194
# Hothorn T and Leisch F 2011. Case studies in reproducibility. Briefings in Bioinformatics 12(3), 288-300.
# Gandrud C 2013 Reproducible Research with R and RStudio. CRC Press Florida.
# Xie Y 2013 Dynamic Documents with R and knitr. CRC Press Florida

## Blogs sourced for this presentation

# http://blog.stodden.net/2013/04/19/what-the-reinhart-rogoff-debacle-really-shows-verifying-empirical-results-needs-to-be-routine/
# http://kbroman.github.io/Tools4RR/
# http://rpubs.com/bbolker/3153
# http://sepwww.stanford.edu/data/media/public/sep//jon/repropreface.html
# http://blog.revolutionanalytics.com/2010/10/a-workflow-for-r.html
# http://www.stanford.edu/~vcs/AAAS2011/
# http://wiki.stodden.net/Best_Practices_for_Researchers_Publishing_Computational_Results
# http://www.reproducibleresearch.net/index.php/RR_links
# http://www.nature.com/nature/focus/reproducibility/
# Claerbout's ranking (http://dx.doi.org/10.1109/5992.881708)
# http://fperez.org/py4science/git.html
# http://ivory.idyll.org/blog/replication-i.html
# http://www.mendeley.com/groups/1142301/reproducible-research/
# http://biostat.mc.vanderbilt.edu/wiki/Main/StatReport
# http://reproducibleresearch.net/index.php/How_to
# http://www.executablepapers.com/index.html
# http://tomwallis.info/category/reproducible-research/
# http://simplystatistics.org/2013/08/21/treading-a-new-path-for-reproducible-research-part-1/
# http://scienceinthesands.blogspot.co.uk/search/label/reproduciblie%20research
# http://scienceinthesands.blogspot.co.uk/2012/08/7-habits-of-open-scientist-2.html
# http://scitation.aip.org/content/aip/journal/cise/11/1/10.1109/MCSE.2009.19
# http://www.reproducibility.org/RSF/book/rsf/scons/paper_html/node2.html
# http://ajrichards.bitbucket.org/lpEdit/ReproducibleResearch.html
# http://kieranhealy.org/blog/archives/2014/01/23/plain-text/
# http://nicercode.github.io/git/
# http://nicercode.github.io/blog/2013-04-05-projects/
# http://ivory.idyll.org/blog/tag/reproducibility.html
# http://yihui.name/en/tags/#Reproducible Research
# http://www.rstudio.com/ide/download/preview
# https://github.com/rstudio/rmarkdown

# web applications, services & organisations 

# http://recomputation.org/
# http://sciencecodemanifesto.org/
# http://researchcompendia.org/
# https://collage.elsevier.com/
# http://www.runmycode.org/home/?/CompanionSite/

## Software cited in the presentation (in order of appearance)

# http://www.r-project.org/
# http://cran.r-project.org/web/packages/roxygen2/index.html
#   http://www.rstudio.com/ide/docs/packages/documentation
#   http://adv-r.had.co.nz/Documenting-functions.html
# http://rcharts.io/
#   http://ramnathv.github.io/rCharts/
#   https://github.com/ramnathv/rCharts
# http://ipython.org/
# https://github.com/att/rcloud
#   http://www.research.att.com/articles/featured_stories/2013_09/201309_SandR.html?fbid=RljYLCmQyyR
# http://daringfireball.net/projects/markdown/
# http://rmarkdown.rstudio.com/
# http://yihui.name/knitr/
#   http://cran.r-project.org/web/packages/knitr/index.html
# http://johnmacfarlane.net/pandoc/
# http://git-scm.com/
#   https://github.com/
#   https://bitbucket.org/
# http://www.rstudio.com/ide/
# http://figshare.com/
# http://datadryad.org/
#

```

