<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />


<meta name="author" content="Brad Boyle" />


<title>TNRS</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/readable.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<script src="site_libs/navigation-1.1/codefolding.js"></script>
<link href="site_libs/highlightjs-1.1/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-1.1/highlight.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>
<link href="site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>

<link rel="stylesheet" href="cm_yeti_bootswatch.css" type="text/css" />

</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
  height: auto;
}
.tabbed-pane {
  padding-top: 12px;
}
button.code-folding-btn:focus {
  outline: none;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 66px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 71px;
  margin-top: -71px;
}

.section h2 {
  padding-top: 71px;
  margin-top: -71px;
}
.section h3 {
  padding-top: 71px;
  margin-top: -71px;
}
.section h4 {
  padding-top: 71px;
  margin-top: -71px;
}
.section h5 {
  padding-top: 71px;
  margin-top: -71px;
}
.section h6 {
  padding-top: 71px;
  margin-top: -71px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>


<div class="container-fluid main-container">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->
<style type="text/css">
.code-folding-btn { margin-bottom: 4px; }
</style>
<script>
$(document).ready(function () {
  window.initializeCodeFolding("show" === "show");
});
</script>




<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}


.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
  padding-left: 25px;
  text-indent: 0;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>

<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-inverse  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">Ecology, Statistics, and Data Science with R</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="Head2_Schedule.html">Schedule</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    R Intro
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="Head1_CourseContent.html">About Course Content</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Module 1: Introduction to R</li>
    <li>
      <a href="00_CourseIntroductionFrame.html">1.0 Course Introduction</a>
    </li>
    <li>
      <a href="01_Rintro.html">1.1 First Steps</a>
    </li>
    <li>
      <a href="02_DataWrangling.html">1.2 Data Wrangling</a>
    </li>
    <li>
      <a href="03_Plotting.html">1.3 Plotting</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Module 2: Spatial Analyses</li>
    <li>
      <a href="04_Spatial.html">2.1 Spatial Data</a>
    </li>
    <li>
      <a href="05_Raster.html">2.2 Spatial Raster Data</a>
    </li>
    <li>
      <a href="06_RasterTwo.html">2.3 More Raster Data</a>
    </li>
    <li class="divider"></li>
    <li class="dropdown-header">Module 3: Reproducible and Interactive Research</li>
    <li>
      <a href="07_Reproducible.html">3.1 Research Reports with R markdown</a>
    </li>
    <li>
      <a href="11_Git.html">3.2 Version control with Github</a>
    </li>
    <li>
      <a href="07_03_RMarkdown_Tools.html">3.3 RMarkdown Tools: Interactive Presentations, Apps, Websites</a>
    </li>
    <li class="divider"></li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Ecoinformatics
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="3_1_Ecoinf_Schedule.html">0 Workshop Schedule</a>
    </li>
    <li>
      <a href="3_8_BIEN_intro.html">1 (BIEN) Botanical Information and Ecology Network</a>
    </li>
    <li>
      <a href="3_7_TNRS.html">2 Taxonomic Name Resolution</a>
    </li>
    <li>
      <a href="3_3_RBIEN_tutorial.html">3 Accessing Plant Biodiversity Data</a>
    </li>
    <li>
      <a href="3_4_wallace.html">4 A streamlined modeling workflow: Wallace</a>
    </li>
    <li>
      <a href="3_5_Minxent.html">5 Integrating biogeographic data </a>
    </li>
    <li>
      <a href="3_2_Networks.html">6 Network Associations</a>
    </li>
    <li>
      <a href="3_6_Teaching_Ecoinformatics.html">7 Teaching Econformatics</a>
    </li>
  </ul>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Ecological Modeling
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="4_1_Wallace_SDMs.html">Range Models with Wallace</a>
    </li>
    <li>
      <a href="20_Intro_Demography.html">Introduction to Demography</a>
    </li>
    <li>
      <a href="21_Intro_IPMs.html">Introduction to Integral Projection Models</a>
    </li>
  </ul>
</li>
<li>
  <a href="Head3_Resources.html">Resources</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/cmerow/RDataScience/tree/gh-pages">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">

<div class="btn-group pull-right">
<button type="button" class="btn btn-default btn-xs dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span>Code</span> <span class="caret"></span></button>
<ul class="dropdown-menu" style="min-width: 50px;">
<li><a id="rmd-show-all-code" href="#">Show All Code</a></li>
<li><a id="rmd-hide-all-code" href="#">Hide All Code</a></li>
</ul>
</div>



<h1 class="title toc-ignore">TNRS</h1>
<h4 class="author"><em>Brad Boyle</em></h4>

</div>


<div>
<object data="3_7_assets/tnrs.pdf" type="application/pdf" width="100%" height="650px">
<p>
It appears you don’t have a PDF plugin for this browser. No biggie… you can <a href="3_7_assets/tnrs.pdf">click here to download the PDF file.</a>
</p>
</object>
</div>
<p>
<a href="3_7_assets/tnrs.pdf">Download the PDF of the presentation</a>
</p>
<p><a href="3_7_TNRS.R"><i class="fa fa-file-code-o fa-3x" aria-hidden="true"></i> The R Script associated with this page is available here</a>. Download this file and open it (or copy-paste into a new script) with RStudio so you can follow along. ** Sorry, this download is not currently working - please follow the code below**</p>
<div id="gui-example" class="section level1">
<h1><span class="header-section-number">1</span> GUI Example</h1>
<pre class="r"><code>###############################################
# Example use of TNRS web interface
# By: Naim Matasci &amp; Brad Boyle (bboyle@email.arizona.edu)
# Date: Jan. 5, 2017
###############################################

library(ape)

#Tree topology from Ackerly, D. 2009. Conservatism and diversification of plant functional traits: Evolutionary rates versus phylogenetic signal. PNAS 106:19699--19706.
lobelioids.string&lt;-&#39;((((((Lobelia_kauaensis,Lobelia_villosa),Lobelia_gloria-montis),(Trematolobelia_kauaiensis,Trematolobelia_macrostachys)),((Lobelia_hypoleuca,Lobelia_yuccoides),Lobelia_niihauensis)),((Brighamia_insignis,Brighamia_rockii),(Delissea_rhytidosperma,Delissea_subcordata))),((((Cyanea_pilosa,Cyanea_acuminata),Cyanea_hirtella),(Cyanea_coriacea,Cyanea_leptostegia)),(((Clermontia_kakeana,Clermontia_parviflora),Clermontia_arborescens),Clermontia_fauriei)));&#39;

#Transform the newick sting into an ape phylo object
tree&lt;-read.tree(text=lobelioids.string)

#Obtain the taxa names
old.names&lt;-tree$tip.label

#Change the underscore characters into blank spaces
old.names&lt;-gsub(&#39;_&#39;,&#39; &#39;,old.names)

# Add an integer identifier to each name
# I&#39;ve turned the names vector into a two column data frame, but you
# could use a matrix if you prefer
old.names.df &lt;- as.data.frame(old.names)
old.names.df[,2] &lt;- old.names.df[,1]
old.names.df[,1] &lt;- seq(1:nrow(old.names.df))
colnames(old.names.df) &lt;- c(&quot;id&quot;,&quot;Name_submitted&quot;)

# Write names and IDs to tab-delimitted file without headers
#  A CSV file would also work
write.table(old.names.df, file=&quot;oldnames.csv&quot;, 
    row.names=FALSE, col.names=FALSE, sep=&quot;,&quot;)
    
#############################################
# STOP HERE AND COMPLETE THE FOLLOWING MANUAL 
# STEPS BEFORE CONTINUING
#
# 1. Upload &amp; process names file with TNRS web interface
# (http://tnrs.iplantcollaborative.org/TNRSapp.html)
#       a. Edit &quot;Name processing settings&quot; on right
#           i. Select sources. 
#               IMPORTANT! If you choose &quot;TPL&quot;, make sure you also 
#               choose ILDIS and GCC. These three together have the same
#               content as The Plant List (www.theplantlist.org)
#       b. Select tab &quot;Upload and Submit List&quot;
#       c. Check box &quot;My file contains an identifier as first column&quot;
#       d. Choose file, enter email, and submit
#
# 2. Download results when ready (you will be notified by email)
#       a. Select tab &quot;Retrieve Results&quot;
#       b. Enter your email and submission key (from notification email)
#       c. Inspect results. When in doubt, use the hyperlinks to research 
#           names at the original source. If you select alternative matches, 
#           these will be downloaded instead as the best match.
#       d. Download the results, with the following options:
#           i. Best matches only
#           ii. Detailed download format
#           iii. UTF-8
#           iv. Rename download file if you wish
#           v. Save the file to this directory
#
# TIP: try using different source to see how they affect you names
#       before and after
#############################################</code></pre>
<pre class="r"><code># Import the TNRS results file as a tab-delimitted text file with headers
# Here we are assuming the default file name supplied by TNRS
tnrs.results.filename &lt;- &quot;tnrs_results.txt&quot;
tnrs.results &lt;- read.table(tnrs.results.filename, header=T, sep=&quot;\t&quot;)

# Merge the old and new names into single data frame, by id
new.names.df &lt;- tnrs.results[, c(&quot;user_id&quot;, &quot;Name_matched&quot;, &quot;Accepted_name&quot;)]
names &lt;- merge(old.names.df, new.names.df, by.x=c(&quot;id&quot;), by.y=c(&quot;user_id&quot;))
names &lt;- data.frame(lapply(names, as.character), stringsAsFactors=FALSE)

# Add a new column to keep clear the distinction between the name matched, 
# the accepted name and the final name used
# Set it to the accepted name to start
names$Final_name &lt;- names$Accepted_name

# If TNRS did not return any accepted name (no match, or name is already accepted), 
# use the final name instead
# WARNING: 
# You might want to research the matched name instead of using it blindly!
names[names[,c(&quot;Accepted_name&quot;)]==&quot;&quot;,c(&quot;Final_name&quot;)] &lt;-
    names[names[,c(&quot;Accepted_name&quot;)]==&quot;&quot;,c(&quot;Name_matched&quot;)] 

# Save the old tree for comparison
tree.old &lt;- tree

#The old taxa names are replaced with the corrected taxa names
tree$tip.label&lt;-names[,c(&quot;Final_name&quot;)]

# Set up mutliplot
par(mfrow=c(1,2))

# Plot the original tree with old names
plot(tree.old, main=&quot;Before name resolution&quot;)

# Plot the tree with resolved names
plot(tree, main=&quot;After name resolution&quot;)</code></pre>
</div>
<div id="api-example" class="section level1">
<h1><span class="header-section-number">2</span> API Example</h1>
<pre class="r"><code>###############################################
# Example use of TNRS api
# By: Naim Matasci
###############################################

library(ape)
library(rjson)  
library(RCurl) 

tnrs.api&lt;-&#39;http://tnrs.iplantc.org/tnrsm-svc&#39;

#Tree topology from Ackerly, D. 2009. Conservatism and diversification of plant functional traits: Evolutionary rates versus phylogenetic signal. PNAS 106:19699--19706.
lobelioids.string&lt;-&#39;((((((Lobelia_kauaensis,Lobelia_villosa),Lobelia_gloria-montis),(Trematolobelia_kauaiensis,Trematolobelia_macrostachys)),((Lobelia_hypoleuca,Lobelia_yuccoides),Lobelia_niihauensis)),((Brighamia_insignis,Brighamia_rockii),(Delissea_rhytidosperma,Delissea_subcordata))),((((Cyanea_pilosa,Cyanea_acuminata),Cyanea_hirtella),(Cyanea_coriacea,Cyanea_leptostegia)),(((Clermontia_kakeana,Clermontia_parviflora),Clermontia_arborescens),Clermontia_fauriei)));&#39;

#Transform the newick sting into an ape phylo object
tree&lt;-read.tree(text=lobelioids.string)

#Obtain the taxa names
old.names&lt;-tree$tip.label

#Change the underscore characters into blank spaces
old.names&lt;-gsub(&#39;_&#39;,&#39; &#39;,old.names)

#Transforms the vector into a string
old.names&lt;-paste(old.names,collapse=&#39;,&#39;)

#The string needs to be URL-encoded
old.names&lt;-curlEscape(old.names)

#Send a request to the TNRS service
url&lt;-paste(tnrs.api,&#39;/matchNames?retrieve=best&amp;names=&#39;,old.names,sep=&#39;&#39;)
tnrs.json&lt;-getURL(url) 

#The response needs to be converted from JSON
tnrs.results&lt;-fromJSON(tnrs.json)

#The corrected names are extracted from the response
names&lt;-sapply(tnrs.results[[1]], function(x) c(x$nameSubmitted,x$acceptedName))
names&lt;-as.data.frame(t(names),stringsAsFactors=FALSE)

#If TNRS did not return any accepted name (no match, or name is already accepted), the submitted name is retained
names[names[,2]==&quot;&quot;,2]&lt;-names[names[,2]==&quot;&quot;,1] 

# Save the old tree so we can compare
tree.old &lt;- tree

#The old taxa names are replaced with the corrected taxa names
tree$tip.label&lt;-names[,2]

# Set up mutliplot
par(mfrow=c(1,2))

# Plot the original tree with old names
plot(tree.old, main=&quot;Before name resolution&quot;)

# Plot the tree with resolved names
plot(tree, main=&quot;After name resolution&quot;)</code></pre>
<p><img src="3_7_TNRS_files/figure-html/unnamed-chunk-4-1.png" /><!-- --></p>
</div>
<div id="diaobolical-example" class="section level1">
<h1><span class="header-section-number">3</span> Diaobolical Example</h1>
<pre class="r"><code>###############################################
# Really diabolical example use of TNRS web interface to scrub
# Al Gentry plot data
#
# By: Brad Boyle (bboyle@email.arizona.edu)
# Date: Jan. 5, 2017
###############################################

# Read in the plot data
# Here we are using the Alwyn Gentry 0.1 ha transect    CERROOLU, from
# cloud forest in Cerro Olumo, Nicaragua, saved as a CSV file. Source:
# http://www.mobot.org/mobot/research/gentry/welcome.shtml
rawdata &lt;- read.csv(&quot;CERROOLU.csv&quot;, header=TRUE)

# Have a look at the data:
rawdata[1:12,] 

# Note the separate columns for FAMILY, GENUS, and SPECIES. Also that 
# SPECIES is actually specific epithet, not the full species name. 
# Note also (1) use of all caps (harder for the TNRS parser
# to separate &amp; classify the name parts), (2) presence of morphospecies (e.g.,
# Desmopsis M1), (3) morphospecies determined to family only (e.g., 
# Acanthaceae M1 M1), and (4) rampant misspellings (trust me)

# Lets pull just the names and add some columns
raw.names &lt;- rawdata[,2:4]
colnames(raw.names) &lt;- c(&quot;famly.orig&quot;, &quot;genus.orig&quot;, &quot;sp.epithet.orig&quot;)

# Form the full species name
raw.names$species.orig &lt;- paste(raw.names$genus.orig, raw.names$sp.epithet.orig, sep=&quot; &quot;)

# Have a look at the first few records. 
raw.names[1:12,]

# Looks like we need to include family to disambiguate morphospecies such 
# as &quot;Acanthaceae M1 M1&quot; which are identified to family only.  We can identify
# these rows by the presence of a numeralspecies.orig
raw.names[grep(&quot;[[:digit:]]&quot;, raw.names$genus.orig), c(&quot;species.orig&quot;)] &lt;- paste(
    raw.names[grep(&quot;[[:digit:]]&quot;, raw.names$genus.orig), 1], 
    raw.names[grep(&quot;[[:digit:]]&quot;, raw.names$genus.orig), 2], 
    raw.names[grep(&quot;[[:digit:]]&quot;, raw.names$genus.orig), 3], 
    sep=&quot; &quot;)
    
# That&#39;s better:
raw.names[1:12,]

# Let&#39;s put just the unique old names in their own data frame
# This step is optional, but if you have a very large number of records it 
# will save TNRS processing time to process unique names only. Also,
# it is easier to inspect the results
unique.names &lt;- unique(raw.names$species.orig)

# We could upload just the names, but let&#39;s add an ID just to be safe
unique.names &lt;- as.data.frame(unique.names)
unique.names[,2] &lt;- unique.names[,1]
unique.names[,1] &lt;- seq(1:nrow(unique.names))
colnames(unique.names) &lt;- c(&quot;id&quot;,&quot;Name_submitted&quot;)

# Write names and IDs to tab-delimitted file without headers
#  A CSV file would also work
write.table(unique.names, file=&quot;unique_names.csv&quot;, 
    row.names=FALSE, col.names=FALSE, sep=&quot;,&quot;)
    
##################################################
# STOP HERE AND PROCESS THE NAMES MANUALLY USING THE TNRS
#
# 1. Upload &amp; process names file with TNRS web interface
# (http://tnrs.iplantcollaborative.org/TNRSapp.html)
#       a. Edit &quot;Name processing settings&quot; on right
#           i. Select source &quot;Tropicos&quot; only. Why? Because Tropicos is
#               the best authority for New World tropical plant names
#       b. Select tab &quot;Upload and Submit List&quot;
#       c. Check box &quot;My file contains an identifier as first column&quot;
#       d. Choose file, enter email, and submit
#
# *** You will be notified by email when results are ready ***
#
# 2. Inspect the results!!! Especially, anything with a flag,. Also select 
#       alternative species matches if applicable. For genera of morphospecies
#       this isn&#39;t usually necessary.
#       a. Select tab &quot;Retrieve Results&quot;
#       b. Enter your email and submission key (from notification email)
#       c. Inspect &amp; research. When in doubt, use the hyperlinks to research 
#           names at the original source. If you select alternative matches, 
#           these will be downloaded instead as the best match. Research 
#           any name with Taxonomic Status &lt;&gt; Accepted or Synonym.
# 
# NOTE 1. The TNRS could not resolve &quot;SARCORHACHIS NARANPANUM&quot;. 
#       Following the link to Tropicos takes you to the page for the genus 
#       &quot;Sarcorhachis&quot;. Click on the link   &quot;Subordinate Taxa&quot;. Scrolling down the 
#       list of species, it appears that the submitted name is an extremely garbled 
#       misspelling of &quot;Sarcorhachis naranjoana&quot;. At this point, you could either fix
#       the name in the existing TNRS results, or change the spelling in the data frame
#       of original names and re-submit to the TNRS. Let&#39;s just fix it ourselves.
#
#   NOTE 2. We have one &quot;Illegitimate&quot; name in our list: &quot;Lasianthea fruticosa&quot;.
#       Tropicos reports that this is a spelling variant of &quot;Lasianthaea fruticosa&quot;, 
#       But for some reason they do not correct it. As a spelling variant, it should be
#       marked Invalid, not Illegitimate. In any case, we&#39;ll fix it later.
#
# NOTE 3: We have one &quot;Invalid&quot; name on our list: &quot;Tetrapteris schiedeana&quot;.
#       This is a spelling variant which was never formally published. Tropicos
#       reports that the name can be &quot;automatically correctable to Tetrapterys&quot;,
#       but for some reason they do not do so. Let&#39;s correct it ourselves.
#
# 3. Download the results, with the following options:
#           a. Best matches only
#           b. Detailed download format
#           c. UTF-8
#           d. Rename download file; let&#39;s call it &quot;unique_names_scrubbed.txt&quot;
#           e. Save the file to this directory
#
##################################################

# Import the TNRS results file as a tab-delimitted text file with headers
# Here we are assuming the default file name supplied by TNRS
tnrs.results.filename &lt;- &quot;unique_names_scrubbed.txt&quot;
tnrs.results &lt;- read.table(tnrs.results.filename, header=T, sep=&quot;\t&quot;)

# Merge the old and new names into single data frame, by id
# Lets include family and unmatched terms, as well as the matched species name.
# Also let&#39;s save authorities. Just because.
new.names.df &lt;- tnrs.results[, c(&quot;user_id&quot;, 
    &quot;Name_matched_accepted_family&quot;, &quot;Name_matched&quot;, 
    &quot;Name_matched_author&quot;, &quot;Name_matched_rank&quot;, 
    &quot;Accepted_name_family&quot;, &quot;Accepted_name&quot;, &quot;Accepted_name_author&quot;,
    &quot;Unmatched_terms&quot;, &quot;Taxonomic_status&quot;
    )]
names &lt;- merge(unique.names, new.names.df, by.x=c(&quot;id&quot;), by.y=c(&quot;user_id&quot;))
names &lt;- data.frame(lapply(names, as.character), stringsAsFactors=FALSE)

# Add some new columns to hold the final family, taxon, morphospecies
# and author. Start by pasting in the accepted names
names$family.final &lt;- names$Accepted_name_family
names$name.final &lt;- names$Accepted_name
names$authority.final &lt;- names$Accepted_name_author
names$morphospecies.final &lt;- &quot;&quot;
names[is.na(names)] &lt;- &quot;&quot;

# Let&#39;s have a look at just the names:
names[ 1:12, c(&quot;Name_submitted&quot;, &quot;Name_matched&quot;, &quot;Accepted_name&quot;, &quot;name.final&quot;)]

# If TNRS did not return any accepted name (no match, or name is already accepted), 
# use the final name instead
names[names[,c(&quot;Accepted_name&quot;)]==&quot;&quot;, c(&quot;family.final&quot;)] &lt;-
    names[names[,c(&quot;Accepted_name&quot;)]==&quot;&quot;, c(&quot;Name_matched_accepted_family&quot;)] 
names[names[,c(&quot;Accepted_name&quot;)]==&quot;&quot;, c(&quot;name.final&quot;)] &lt;-
    names[names[,c(&quot;Accepted_name&quot;)]==&quot;&quot;,c(&quot;Name_matched&quot;)] 
names[names[,c(&quot;Accepted_name&quot;)]==&quot;&quot;, c(&quot;authority.final&quot;)] &lt;-
    names[names[,c(&quot;Accepted_name&quot;)]==&quot;&quot;,c(&quot;Name_matched_author&quot;)] 

# Check again
names[ 1:12, c(&quot;Name_submitted&quot;, &quot;Name_matched&quot;, &quot;Accepted_name&quot;, &quot;name.final&quot;)]

# Let get rid of authority for names only identified to family or genus
names[names[, c(&quot;Name_matched_rank&quot;)]==&quot;family&quot; ,c(&quot;authority&quot;)] &lt;- &quot;&quot;
names[names[, c(&quot;Name_matched_rank&quot;)]==&quot;genus&quot; ,c(&quot;authority&quot;)] &lt;- &quot;&quot;

# Fix the three invalid/illegitimate or unmatched names:
names[names[, c(&quot;Name_submitted&quot;)]==&quot;LASIANTHEA FRUTICOSA&quot; , c(&quot;family.final&quot;)] &lt;- &quot;Asteraceae&quot;
names[names[, c(&quot;Name_submitted&quot;)]==&quot;LASIANTHEA FRUTICOSA&quot; , c(&quot;name.final&quot;)] &lt;- &quot;Lasianthaea fruticosa&quot;
names[names[, c(&quot;Name_submitted&quot;)]==&quot;LASIANTHEA FRUTICOSA&quot; , c(&quot;authority.final&quot;)] &lt;- &quot;(L.) K.M. Becker&quot;

names[names[, c(&quot;Name_submitted&quot;)]==&quot;SARCORHACHIS NARANPANUM&quot; , c(&quot;family.final&quot;)] &lt;- &quot;Piperaceae&quot;
names[names[, c(&quot;Name_submitted&quot;)]==&quot;SARCORHACHIS NARANPANUM&quot; , c(&quot;name.final&quot;)] &lt;- &quot;Sarcorhachis naranjoana&quot;
names[names[, c(&quot;Name_submitted&quot;)]==&quot;SARCORHACHIS NARANPANUM&quot; , c(&quot;authority.final&quot;)] &lt;- &quot;(C. DC.) Trel.&quot;

names[names[, c(&quot;Name_submitted&quot;)]==&quot;TETRAPTERIS SCHIEDEANA&quot; , c(&quot;family.final&quot;)] &lt;- &quot;Malpighiaceae&quot;
names[names[, c(&quot;Name_submitted&quot;)]==&quot;TETRAPTERIS SCHIEDEANA&quot; , c(&quot;name.final&quot;)] &lt;- &quot;Tetrapterys schiedeana&quot;
names[names[, c(&quot;Name_submitted&quot;)]==&quot;TETRAPTERIS SCHIEDEANA&quot; , c(&quot;authority.final&quot;)] &lt;- &quot;Schltdl. &amp; Cham.&quot;

# Finally, let&#39;s put morphospecies back together:
names[is.na(names)] &lt;- &quot;&quot;
names$morphospecies.final &lt;- paste(names$name.final, names$Unmatched_terms, sep=&quot; &quot;)
# Trim whitespace. You may need to use the fancier version if you have R earlier than 3.2.0
names$morphospecies.final &lt;- trimws(names$morphospecies.final)

# Check the final result:
names[ , c(&quot;Name_submitted&quot;, &quot;Name_matched_accepted_family&quot;, &quot;Name_matched&quot;, &quot;Accepted_name_family&quot;, &quot;Accepted_name&quot;, 
    &quot;family.final&quot;, &quot;name.final&quot;, &quot;morphospecies.final&quot;
    )]
    
# Pretty good, but not quite perfect. Note that &quot;real&quot; specific epithets that did not 
# match (as  opposed to morphospecies names) weren&#39;t dumped to Unmatched_terms.
# So you&#39;ll still need to do some special handling to reform a morphospecies for such
# name. (E.g., &quot;STEMMADENIA DONNELL-SMITHII&quot;; try extracting the unmatched part
# and append to the matched genus: &quot;Stemmadenia DONNELL-SMITHII&quot;). And you&#39;ll 
# need to remove the message &quot;No suitable matches found.&quot; that the TNRS pastes
# into complete non-matches

# Final step: I&#39;ll leave it to you join your final names back to the original 
# data file. </code></pre>
</div>


<!-- give the footer some space -->
<br/>
<br/>

<footer id="site-footer">
  <div id="footer1">
  <a href="https://cmerow.github.io"><img src="img/cory.png" alt="logo" width=40px></a>
  <!--
  <a href="http://adamwilson.us/#contact"><i class="fa fa-envelope fa-2x"></i></a> 
  <a href="https://twitter.com/AdamWilsonLab"><i class="fa fa-twitter fa-2x"></i></a> 
  <a href="https://github.com/AdamMWilson"><i class="fa fa-github fa-2x"></i></a>
  -->
  </div>
  <div id="footer2">
  <a rel="license" property="http://creativecommons.org/ns#license"
  href="http://creativecommons.org/licenses/by/4.0/" ><img src="img/cc-by.svg" alt="cc-by"/></a> 
  </div>
</footer>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>


</body>
</html>
